org.opent2t.sample.hub.superpopular/com.hue.hub/js/thingTranslator.js

```diff
@@ -19,39 +19,170 @@ class Translator {
+        }
+
+        // this comes from the onboardFlow property 
+        // as part of the schema and manifest.xml
+        var options = {
+            url: "https://api.meethue.com/oauth2/refresh?&grant_type=refresh_token",
+            method: "POST",
+            headers: {
+                'cache-control': 'no-cache'
+            },
+            followAllRedirects: true,
+        };
+
+        return request(options)
+            .then(() => {
+                //Do Nothing
+            })
+            .catch((err) => {
+                if (err.statusCode == '401') {
+                    //extract nonce code from header
+                    var digestHeader = err.response.headers['www-authenticate'];
+                    var nonce = digestHeader.substr(digestHeader.indexOf('nonce=\"') + 7, 32);
+
+                    //Compute digest header response
+                    var disgestHeaderContent = 'username=\"' + authInfo[0].client_id + '\", ';
+                    disgestHeaderContent += 'realm=\"oauth2_client@api.meethue.com\", ';
+                    disgestHeaderContent += 'nonce=\"' + nonce + '\", ';
+                    disgestHeaderContent += 'uri=\"/oauth2/refresh\", ';
+                    var HASH1 = crypto.createHash('md5').update(authInfo[0].client_id + ':oauth2_client@api.meethue.com:' + authInfo[0].client_secret).digest('hex');
+                    var HASH2 = crypto.createHash('md5').update('POST:/oauth2/refresh').digest('hex');
+                    var authHeaderResponse = crypto.createHash('md5').update(HASH1 + ':' + nonce + ':' + HASH2).digest('hex');
+                    disgestHeaderContent += 'response=\"' + authHeaderResponse + '\"';
+
+                    options.headers = {
+                        'Accept': 'application/json',
+                        'Authorization': 'Digest ' + disgestHeaderContent,
+                        'Content-type': 'application/x-www-form-urlencoded'
+                    };
+
+                    options.body = 'refresh_token=' + this._accessToken.refreshToken;
+                    return request(options)
+                        .then((body) => {
+                            var tokenInfo = JSON.parse(body); // This includes refresh token, scope etc..
+                            return new accessTokenInfo(
+                                        tokenInfo.access_token,
+                                        this._add2CurrentUTC(tokenInfo.access_token_expires_in),
+                                        tokenInfo.refresh_token,
+                                        this._add2CurrentUTC(tokenInfo.refresh_token_expires_in),
+                                        tokenInfo.token_type,
+                                        this._accessToken.bridgeId,
+                                        this._accessToken.whitelistId
+                                    );
+                        }).catch(function (err) {
```

Comment:
I think this part since huge because I broke down the digest authentication part for easier code reading. I can write an helper function for digest authentication if needed.

---
In reply to: [94458370](https://github.com/openT2T/translators/pull/143#discussion_r94458370) [](ancestors = 94458370)